File: OtherMethods/fdaimage.py
import numpy as np

import rpy2.robjects as robjects
import rpy2.robjects.numpy2ri

from utils import erint


rpy2.robjects.numpy2ri.activate()
r = robjects.r
r['source']('fdaimage.R')


def get_coordinates(dim):
    d = np.prod(dim)
    coords = np.c_[np.unravel_index(np.arange(d).reshape(-1, 1), dim)]
    coords = coords / coords.max(0, keepdims=True)
    return coords


def fdaimage(
        Y, X, Z, alpha=0.05, nboot=10,
        boundary='rectangle', return_sigma=True, verbose=False):

    assert isinstance(Y, np.ndarray)
    assert isinstance(X, np.ndarray)
    assert Y.ndim == 2
    assert X.ndim == 2
    assert Z.ndim == 2
    assert X.shape[0] == Y.shape[0]
    assert Y.shape[1] == Z.shape[0]
    assert Z.shape[1] == 2

    out = r['fdaimage'](
            Y, X, Z, boundary=boundary, alpha=alpha, nboot=nboot,
            return_sigma=return_sigma, verbose=verbose, debug=False)
    beta, sigma, ci_lower, ci_upper, alpha_adj = out
    beta = np.array(beta).T
    sigma = np.array(sigma).T
    ci_lower = np.array(ci_lower).T
    ci_upper = np.array(ci_upper).T
    return beta, sigma, ci_lower, ci_upper, alpha_adj


def fdaimage3d(
        Y, X, Z, boundary='rectangle', return_sigma=True, verbose=False):

    assert isinstance(Y, np.ndarray)
    assert isinstance(X, np.ndarray)
    assert isinstance(Z, np.ndarray)
    assert Y.ndim == 2
    assert X.ndim == 2
    assert Z.ndim == 2
    assert X.shape[0] == Y.shape[0]
    assert Y.shape[1] == Z.shape[0]
    assert Z.shape[1] == 3

    beta = np.zeros((X.shape[1], Y.shape[1]))
    radi = np.zeros_like(beta)
    idxs, ys, xs, zs = [], [], [], []
    for w in np.unique(Z[:, 2]):
        idx = Z[:, 2] == w
        idxs.append(idx)
        ys.append(Y[:, idx])
        xs.append(X)
        zs.append(Z[idx, :2])

    for idx, y, x, z in zip(idxs, ys, xs, zs):
        if verbose:
            erint(Z[idx][0])
        b, r = fdaimage(
                y, x, z, boundary=boundary,
                return_sigma=return_sigma, verbose=verbose)
        beta[:, idx] = b
        if np.isfinite(r).all() and np.min(r) >= 0:
            radi[:, idx] = r
        else:
            radi[:, idx] = np.nan

    # from itertools import starmap
    # bs = list(starmap(fdaimage, zip(ys, xs, zs)))

    # for idx, b in zip(idxs, bs):
    #     beta[:, idx] = b

    return beta, radi


def test_fdaimage():
    V = (16, 16)
    N = 50
    Q = 3
    Y = np.random.randn(N, np.prod(V))
    X = np.random.randn(N, Q)
    Z = get_coordinates(V)
    beta, radi = fdaimage(Y, X, Z)
    assert beta.shape == (Q, np.prod(V))
    assert radi.shape == (Q, np.prod(V))


def test_fdaimage3d():
    V = (16, 16, 4)
    N = 50
    Q = 3
    Y = np.random.randn(N, np.prod(V))
    X = np.random.randn(N, Q)
    Z = get_coordinates(V)
    beta, radi = fdaimage3d(Y, X, Z)
    assert beta.shape == (Q, np.prod(V))
    assert radi.shape == (Q, np.prod(V))


def main():
    # test_fdaimage()
    test_fdaimage3d()


if __name__ == '__main__':
    main()

File: OtherMethods/fdaimage.R
#! /usr/bin/env Rscript
rm(list=ls())

library(FDAimage)

# Y: image
# X: covariate
# Z: coordinates (2D only)
# V: vertices
# Tr: triangles (list of triplets of vertices)
fdaimage = function(Y, X, Z, V=NULL, Tr=NULL, boundary="chull",
         d=5, r=1, nfold=10, lambda=10^(seq(-6,6,by=1)),
         nboot=100, alpha0=0.05, return_sigma=F, verbose=F, debug=F){
    if(debug){
        save(Y, X, Z, file="tmp.RData")
    }
    n = dim(Y)[1]
    N = dim(Y)[2]
    if(is.null(V) | is.null(Tr)){
        c1 = 0.5
        nT = floor(min(c1 * n^(1/(2*d+2)) * N^(1/2), N / 10))
        nT = max(8, nT)
        nMesh = floor(sqrt(nT/2))
        if(boundary == "rectangle"){
            umin = min(Z[,1])
            umax = max(Z[,1])
            vmin = min(Z[,2])
            vmax = max(Z[,2])
            us = seq(umin, umax, length.out=nMesh+1)
            vs = seq(vmin, vmax, length.out=nMesh+1)
            boundary = rbind(cbind(umin, vs), cbind(umax, vs), cbind(us, vmin), cbind(us, vmax))
        } else if(boundary == "chull"){
            # use the convex hull of all the points
            boundary = Z[chull(Z),]
        }
        #pngfile = ifelse(debug, 'tmp.png', './dev/null')
        #png(pngfile)
        VTr = TriMesh(boundary, nMesh)
        #dev.off()
        V = VTr$V 
        Tr = VTr$Tr
        if(verbose){
            print(nT)
            print(nMesh)
            print(dim(boundary))
            print(dim(V))
            print(dim(Tr))
        }
    }
    cv = cv.FDAimage(Y, X, Z, V, Tr, d, r, lambda, nfold)
    lamc = cv$lamc
    if(return_sigma){
        fdaout = scc.FDAimage(Y, X, Z, V, Tr, d, r, lamc, alpha0, nboot)
        beta = fdaout$beta.hat
        sigma = fdaout$Sigma
        ci_lower = fdaout$cc.l
        ci_upper = fdaout$cc.u
        alpha_adj = fdaout$alpha.adj
    } else {
        fdaout = fit.FDAimage(Y, X, Z, V, Tr, d, r, lamc)
        beta = fdaout$beta
        sigma = NA
        ci_lower = NA
        ci_upper = NA
        alpha_adj = NA
    }
    out = list(
        beta=beta, sigma=sigma,
        ci_lower=ci_lower, ci_upper=ci_upper,
        alpha_adj=alpha_adj)
    return(out)
}
